crowbar_framework/app/models/crowbar_service.rb

```diff
@@ -32,6 +32,73 @@ def role_constraints
     end
   end
 
+  def shutdown_openstack_at_nodes
+    db_nodes = []
+    NodeObject.find("state:crowbar_upgrade").each do |node|
+
+      # Find the nodes with Database role
+      # (we can't look at node.roles, because that list has been reduced to just crowbar-upgrade)
+      node.crowbar["run_list_map"].each do |role, v|
+        db_nodes.push node.name if role == "database-server"
+      end
+
+      # mark the position in the upgrade process
+      node["crowbar_wall"]["crowbar_openstack_shutdown"] = true
+      node.save
+    end
+
+    # Adapt current proposal, so only non-db nodes have crowbar-upgrade role
+    # (crowbar-upgrade role is currently assigned to all nodes)
+    proposal = ProposalObject.find_proposal("crowbar", "default")
+    proposal["deployment"]["crowbar"]["elements"]["crowbar-upgrade"] -= db_nodes
+    proposal.save
+
+    return true if proposal["deployment"]["crowbar"]["elements"]["crowbar-upgrade"].empty?
+
+    # Commit proposal so the shutdown actions from crowbar-upgrade get executed for non-db nodes
+    proposal_commit("default", false, false)
+  end
+
+  def dump_database
+    # Find the nodes with Database role
+    # (we can't look at node.roles, because that list has been reduced to just crowbar-upgrade item)
+    db_nodes = NodeObject.find("state:crowbar_upgrade AND run_list_map:database-server")
+
```

Comment:
Ok, but shouldn't we react in someway if ```dump_database``` is called and non of the nodes have the right state? Doesn't it indicate that something went wrong? It might be a good idea to tell the user about that.

BTW,  ```dump_database``` is also a pretty ambiguous name here. Can we change this to something that is indicating that we're dumping the openstack database on the client nodes? (sorry, once more)